contracts/utils/version-control/ConditionalVersionControl.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { ERC1967Upgrade } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol";
+import { IConditionalVersionControl } from "../../interfaces/utils/IConditionalVersionControl.sol";
+import { ErrorHandler } from "../../libraries/ErrorHandler.sol";
+import { AddressArrayUtils } from "../../libraries/AddressArrayUtils.sol";
+import { ErrOnlySelfCall } from "../CommonErrors.sol";
+
+/**
+ * @title ConditionalVersionControl
+ * @dev A contract that allows conditional version control of contract implementations.
+ */
+abstract contract ConditionalVersionControl is IConditionalVersionControl, ERC1967Upgrade {
+  using ErrorHandler for bool;
+  using AddressArrayUtils for address[];
+
+  error ErrInvalidArguments(address proxyStorage, address currentVersion, address newVersion);
+  /// @dev Error of set to non-contract.
+  error ErrZeroCodeContract(address addr);
+  /// @dev Error when contract which delegate to this contract is not compatible with ERC1967
+  error ErrDelegateFromUnknownOrigin(address addr);
+
+  address internal immutable _proxyStorage;
+  address internal immutable _newVersion;
+  address internal immutable _currentVersion;
+
+  modifier onlySelfCall() {
+    _requireSelfCall();
+    _;
+  }
+
+  modifier whenConditionsAreMet() virtual;
+
+  /**
+   * @dev Modifier that only allows delegate calls from admin proxy storage.
+   */
+  modifier onlyDelegateFromProxyStorage() virtual {
+    _requireDelegateFromProxyStorage();
+    _;
+  }
+
+  /**
+   * @dev Modifier that only allows contracts with code.
+   * @param addr The address of the contract to check.
+   */
+  modifier onlyContract(address addr) {
+    _requireHasCode(addr);
+    _;
+  }
```

Comment:
Not a change request but I see lots of this implementation in this contract. Is there any way to reuse these implementations without rewriting them?